apps/arweave/src/ar_merkle.erl

```diff
@@ -190,6 +196,112 @@ validate_path_strict_data_split(_ID, _Dest, _LeftBound, _RightBound, _Path, _Pat
+validate_path2(ID, Dest, RightBound, Path) when Dest < 0 ->
+	validate_path2(ID, 0, RightBound, Path);
+validate_path2(ID, Dest, RightBound, Path) ->
+	validate_path2(ID, Dest, 0, RightBound, Path, byte_size(Path), RightBound, 0).
+
+validate_path2(ID, _Dest, LeftBound, RightBound,
+		<< Data:?HASH_SIZE/binary, EndOffset:(?NOTE_SIZE*8) >>, PathSize, DataSize,
+		LeftBoundShift) ->
+	case EndOffset - LeftBound > ?DATA_CHUNK_SIZE
+			orelse RightBound - LeftBound > ?DATA_CHUNK_SIZE of
+		true ->
+			io:format("Bounds: E=~B L=~B R=~B~n", [EndOffset, LeftBound, RightBound]), % TODO
+			false;
+		false ->
+			ChunkSize = EndOffset - LeftBound,
+			ValidateSplit =
+				case validate_strict_split of
+					_ when ChunkSize == (?DATA_CHUNK_SIZE) ->
+						LeftBound rem (?DATA_CHUNK_SIZE) == 0;
+					_ when EndOffset == DataSize ->
+						Border = RightBound - RightBound rem (?DATA_CHUNK_SIZE),
+						RightBound rem (?DATA_CHUNK_SIZE) > 0
+								andalso LeftBound =< Border;
+					_ when PathSize > ChunkSize ->
+						false;
+					_ ->
+						LeftBound rem (?DATA_CHUNK_SIZE) == 0
+								andalso DataSize - LeftBound > (?DATA_CHUNK_SIZE)
+								andalso DataSize - LeftBound < 2 * (?DATA_CHUNK_SIZE)
+				end,
+			case ValidateSplit of
+				false ->
+					io:format("Split: E=~B L=~B R=~B C=~B~n",
+							[EndOffset, LeftBound, RightBound, ChunkSize]), % TODO
+					false;
+				true ->
+					case hash([hash(Data), hash(note_to_binary(EndOffset))]) of
+						ID ->
+							{Data, LeftBoundShift + LeftBound,
+								LeftBoundShift + max(min(RightBound, EndOffset),
+											LeftBound + 1)};
+						_ ->
+							io:format("Hashing: E=~B L=~B R=~B ID: ~s~n",
+									[EndOffset, LeftBound, RightBound,
+									ar_util:encode(ID)]), % TODO
+							false
+					end
+			end
+	end;
+validate_path2(ID, Dest, LeftBound, RightBound,
+		<< 0:(?HASH_SIZE*8), L:?HASH_SIZE/binary, R:?HASH_SIZE/binary,
+			Note:(?NOTE_SIZE*8), Rest/binary >>, PathSize, DataSize,
+			LeftBoundShift) ->
+	case hash([hash(L), hash(R), hash(note_to_binary(Note))]) of
+		ID ->
+			{Path, NextLeftBound, NextRightBound, Dest2, NextLeftBoundShift} =
+				case Dest < Note of
+					true ->
+						Note2 = min(RightBound, Note),
+						{L, 0, Note2 - LeftBound, Dest - LeftBound,
+								LeftBoundShift + LeftBound};
+					false ->
+						Note2 = max(LeftBound, Note),
+						{R, 0, RightBound - Note2,
+								Dest - Note2,
+								LeftBoundShift + Note2}
+				end,
+			io:format("Passing: E=~B L=~B R=~B Shift: ~B NextShift: ~B~n",
+					[Dest, LeftBound, RightBound, LeftBoundShift, NextLeftBoundShift]), % TODO
+			validate_path2(Path, Dest2, NextLeftBound, NextRightBound, Rest,
+					PathSize, DataSize, NextLeftBoundShift);
+		_ ->
+			false
+	end;
+validate_path2(ID, Dest, LeftBound, RightBound,
+		<< L:?HASH_SIZE/binary, R:?HASH_SIZE/binary, Note:(?NOTE_SIZE*8), Rest/binary >>,
+		PathSize, DataSize, LeftBoundShift) ->
+	case hash([hash(L), hash(R), hash(note_to_binary(Note))]) of
```

Comment:
can this be combined with the `validate_path` code [here](https://github.com/ArweaveTeam/arweave/pull/433/files#diff-da5e1b73bd29394af0e871a3dc4b2299cafb3a0cc8aeb5b072f362667865918aL59-L71)?